File: telegram-bot-bumblebee-core/src/main/java/com/github/bumblebee/command/autocomplete/AutocompleteHandler.java
package com.github.bumblebee.command.autocomplete;

import com.github.bumblebee.command.ChainedMessageListener;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import telegram.api.BotApi;
import telegram.domain.Update;

import java.util.HashMap;
import java.util.Map;
import java.util.regex.Pattern;


@Component
public class AutocompleteHandler extends ChainedMessageListener {

    private static final Logger log = LoggerFactory.getLogger(AutocompleteHandler.class);

    private final BotApi botApi;
    private final Map<String, String[]> autocompletes = new HashMap<>();
    private final String pattern;

    @Autowired
    public AutocompleteHandler(BotApi botApi, AutocompleteConfig config) {
        this.botApi = botApi;
        this.pattern = Pattern.compile("[^\\/\\n]+\\/([^\\/\\n]+\\/)*[^\\/\\n]+").toString();
        config.getTemplates().forEach(this::addTemplate);
    }

    public boolean addTemplate(String phraseTemplate) {
        if (isValidTemplate(phraseTemplate)) {
            String patternKey = phraseTemplate.substring(0, phraseTemplate.indexOf('/'));
            String[] patternValue = phraseTemplate.replaceFirst(Pattern.quote(patternKey + "/"), "").split("/");
            autocompletes.put(patternKey, patternValue);
            return true;
        }
        return false;
    }

    private boolean isValidTemplate(String argument) {
        if (!argument.matches(pattern)) {
            return false;
        }
        return true;
    }


    @Override
    public boolean onMessage(Long chatId, String message, Update update) {
        try {
            if (autocompletes.containsKey(message)) {
                for (String text : autocompletes.get(message)) {
                    if (text.startsWith("stickerId:")) {
                        botApi.sendSticker(chatId, text.replace("stickerId:", ""));
                    } else {
                        botApi.sendMessage(chatId, text);
                    }
                    Thread.sleep(400);
                }
                return true;
            }
        } catch (InterruptedException e) {
            log.warn("Failed to sleep", e);
        }
        return false;
    }
}

File: telegram-bot-bumblebee-core/src/main/kotlin/com/github/bumblebee/command/autocomplete/AutoCompleteHandler.kt
package com.github.bumblebee.command.autocomplete

import com.github.bumblebee.command.ChainedMessageListener
import com.github.bumblebee.command.autocomplete.entity.AutoCompletePhrase
import com.github.bumblebee.command.autocomplete.service.AutoCompleteService
import com.github.telegram.api.BotApi
import com.github.telegram.domain.Update
import org.springframework.stereotype.Component
import java.util.concurrent.ConcurrentHashMap


@Component
class AutoCompleteHandler(
    private val botApi: BotApi,
    private val service: AutoCompleteService
) : ChainedMessageListener() {

    private val completions: MutableMap<String, List<String>> = service.fetchCompletions().associateByTo(
        destination = ConcurrentHashMap(),
        keySelector = { it.phraseKey },
        valueTransform = { splitPhrases(it.phrasePattern) }
    )

    fun addCompletion(template: String): Boolean {
        if (!isValidTemplate(template)) {
            return false
        }

        val parts = splitPhrases(template, limit = 2)
        val patternKey = parts.first()
        val patternPhrase = parts.last()

        if (completions.containsKey(patternKey))
            service.updateAutoCompletePhrase(patternKey, patternPhrase)
        else
            service.storeAutoCompletePhrase(AutoCompletePhrase(patternKey, patternPhrase))
        completions[patternKey] = splitPhrases(patternPhrase)
        return true
    }

    private fun isValidTemplate(argument: String): Boolean = templateRegex.matches(argument)

    private fun splitPhrases(pattern: String, limit: Int = 0) = pattern.split('/', limit = limit)

    override fun onMessage(chatId: Long, message: String?, update: Update): Boolean {
        message?.let {
            completions[it]?.let { phrases ->
                phrases.forEach { phrase ->
                    if (phrase.startsWith(STICKER_PREFIX)) {
                        botApi.sendSticker(chatId, phrase.removePrefix(STICKER_PREFIX))
                    } else {
                        botApi.sendMessage(chatId, phrase)
                    }
                    Thread.sleep(400)
                }
                return true
            }
        }
        return false
    }

    fun availableCompletions(): Set<String> = completions.keys

    companion object {
        private const val STICKER_PREFIX = "stickerId:"
        private val templateRegex = """[^/\n]+/([^/\n]+/)*[^/\n]+""".toRegex()
    }
}


